feat: handle complete profile form submission

sounmind opened the pull request:
Co-authored-by: Manvinderjit Rupal <manvinderjit@users.noreply.github.com>## Summary

Add the Complete Profile form submission function

## Tests

Tested manually. Pending unit, integration, e2e tests.

## Additional information

User must be logged in to access the page.

sounmind committed:
fix: change linkedin social icon to mastodon

Co-authored-by: Manvinderjit Rupal <manvinderjit@users.noreply.github.com>

sounmind committed:
feat: handle complete profile form submission

Co-authored-by: Manvinderjit <91133409+manvinderjit@users.noreply.github.com>

manvinderjit committed:
feat: add complete profile form submission

madcampos commented:
@manvinderjit what is the thing with cookies?

madcampos commented on src/utilities/getApiUrl.ts:
```diff
@@ -1,3 +1,3 @@
 export function getApiUrl() {
-	return import.meta.env['APP_API_URL'] ?? 'https://community-hub.torontojs.com/';
+	return import.meta.env['APP_API_URL'] as string ?? 'https://community-hub.torontojs.com/';
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** The null check (`??`) should already handle the case where the variable doesn't exist., so we need the extra type casting?

Env vars are always `string`s or `undefined`. The extra case we could handle is an empty string

manvinderjit replied:
This is outdated, the file is now removed.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -16,30 +17,98 @@ interface SocialIcons {
+	return regex.test(birthday);
+};
+
+const isRealDate = (birthday: string): boolean => {
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏
```suggestion
/**
 * @deprecated We will remove update this when Temporal becomes stable
 */
const isRealDate = (birthday: string): boolean => {
```

I think this will change once we have better support for Temporal, so let's mark it as such.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -16,30 +17,98 @@ interface SocialIcons {
 }
 
+// Validation helpers
+const isValidBirthdayFormat = (birthday: string): boolean => {
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** I think this function would benefit from some documentation...

Documentation for Typescript and Javascript follow the [JSDoc](https://jsdoc.app/) format.
In [TS](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) case it can be simplified as the types for parameters and function return are already provided by the language.

```suggestion
/**
 * <Insert docs here>
 */
const isValidBirthdayFormat = (birthday: string): boolean => {
```

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -16,30 +17,98 @@ interface SocialIcons {
+	return date.getMonth() + 1 === month && date.getDate() === day;
+};
+
+const updateProfile = async (args: UpdateProfileParams, profileId: string) => {
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏
```suggestion
const updateProfile = async (data: UpdateProfileParams, profileId: string) => {
```

This is the data to be processed, so it is better to name it `data` or `body`, or other descriptive name.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -16,30 +17,98 @@ interface SocialIcons {
+			body: JSON.stringify({ ...args }),
+			credentials: 'include'
+		});
+		if (!response.ok) {
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Please confirm the status code for this request, if it is not a 200 - OK, this will fail.‏

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -16,30 +17,98 @@ interface SocialIcons {
+		});
+		if (!response.ok) {
+			const errorData = await response.json();
+			console.log('Response not ok: ', errorData);
```
[Emo log](https://www.youtube.com/watch?v=dhZTNgAs4Fc)

manvinderjit replied:
Console.log is present because we are waiting on design updates for error message component. Will update this once we have that available.

madcampos replied:
Was just a silly comment to link to an emo music video 🙊

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -16,30 +17,98 @@ interface SocialIcons {
+		{ name: 'Mastodon', element: <Mastodon />, visible: true },
 		{ name: 'BlueSky', element: <BlueSky />, visible: true },
-		{ name: 'X', element: <XTwitter />, visible: true },
+		{ name: 'Twitter', element: <XTwitter />, visible: true },
```
```suggestion
		{ name: 'X/Twitter', element: <XTwitter />, visible: true },
```

manvinderjit replied:
![image](https://github.com/user-attachments/assets/ba615bd9-3f35-423e-adc6-01a851a47a45)

Was keeping this aligned with the backend enum as shown in the image.
1. Does zod enum allow `X/Twitter`? 
2. Do you recommend using lower case for such enums, like platform name?

madcampos replied:
Ah, on this case, yes. Let's keep two things:
- "id", lowercase, agnostic name
- "display name", can be whatever, have formatting, used for display

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -57,11 +126,16 @@ const CompleteProfile = () => {
-			const skillId = skillName + Date.now();
-			setSkills([...skills, { id: skillId, name: skillName }]);
-			if (skillInputRef.current) { skillInputRef.current.value = ''; }
+		const index = profileData.skills.findIndex((skill) => skill.trim().toLocaleLowerCase() === skillName.trim().toLocaleLowerCase());
+		if (skillName.trim() !== '' && index == -1) {
+			setProfileData((prev) => {
+				return {
+					...prev,
+					skills: [...prev.skills, skillName]
+				};
+			});
 		}
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** ‏
```suggestion
		setProfileData((prev) => {
			return {
				...prev,
				skills: [...new Set([...prev.skills, skillName.toLowerCase()])]
			};
		});
```

A [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) guarantees each item is unique. A neat trick to filter an array of primitive values is to convert it to a set and then back to an array like so:

```typescript
[...new Set(originalArray)]
```

The Set constructor receives an iterable, like an array; Then we use the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) to iterate over all items of the set; and lastly wrap it in an array literal to convert back to an array.

madcampos replied:
Side note: let's save all skills as lowercase strings to make it easier to compare

madcampos replied:
Side note: only use locale aware conversions if you are in control of the locale being used. If you let the browser do it, unexpected results may occur as the case rules for different locales are different.

manvinderjit replied:
Good to know about the usage of set. Thanks for sharing.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
+	const handleLinksInputChange = (platform: string, value: string) => {
+		const trimmedValue = value.trim();
+
+		setProfileData((prev) => {
+			const existingLink = prev.links.find((link) => link.platform === platform);
+
+			let updatedLinks;
+			if (existingLink) {
+				if (trimmedValue === '') {
+					updatedLinks = prev.links.filter((link) => link.platform !== platform);
+				} else {
+					updatedLinks = prev.links.map((link) => link.platform === platform ? { ...link, url: trimmedValue } : link);
+				}
+			} else if (trimmedValue !== '') {
+				updatedLinks = [...prev.links, { platform, url: trimmedValue }];
+			} else {
+				return prev;
+			}
+
+			return {
+				...prev,
+				links: updatedLinks
+			};
+		});
```
```suggestion
		if (trimmedValue) {
			setProfileData((prev) => {
				const existingLinkIndex = prev.links.findIndex((link) => link.platform === platform);
				const updatedLinks = prev.links;
				
				if (existingLinkIndex !== -1) {
					updatedLinks.splice(existingLinkIndex, 1);
				}

				return {
					...prev,
					links: [...updatedLinks, {
						platform,
						value: trimmedValue
					}]
				};
			});
		}
```

1. Only update the value if there is anything to update
2. Using [`Array.findIndex`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex) and [`Array.splice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) will remove the item from the array while preserving the order of items if needed. (Another option is to just filter the item out).
3. The spread operator + the item is an easy shortcut to appending items to an array. Don't abuse it though as it will iterate over all items and then append the new one.

manvinderjit replied:
@madcampos I am updating the value when its empty because I want to remove the value if a user empties out the input field after initially typing. Otherwise the input will look like { platform: "LinkedIn", url: "" }.
This removes the respective object from the links array altogether.

madcampos replied:
That is okay to have a sparse array of items. But then again: if we have a defined set of items, why not save them as an object with the format:
```typescript
Record<PlatformName = string, Url = string>
```

manvinderjit replied:
We are not manipulating this state anymore with onChange.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
+		}));
+	};
+
+	const handleSubmit: FormEventHandler<HTMLFormElement> = async (event) => {
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** If we are already handling the form's `submit` event, in theory we don't have to handle any other `input` or `change` events.‏

I figure we still have to do it though because react is a controling and gaslighting little bastard that freaks out if things change without it knowing.

So yeah, two options here:
- Be controlling and have a state variable for every single thing.
- Add `ref`s to every input so react doesn't destroy them and leave us alone.

madcampos replied:
Curveball: when we wrap the inputs into a component, we should keep `ref`s to them. That way we have less work to do regarding react (i.e. we won't need to have state for every single piece of data) and form submitions will be a matter of creating a new `FormData` for the form element...

The only odd thing here will be "custom" inputs like the skills list or birthday. But that can be handled on a case by case basis.

manvinderjit replied:
The Complete Profile is a really complex page
1. We are fetching some data from the backend, like email, name.
2. We are adding location data with radio buttons, aka sliders.
3. We are adding skills data , which is an array of strings.
4. We are adding links, which are an object that must follow an enum key and a string value pattern.
5. The page will have two separate forms, one for updating profile, and another for updating image.

Therefore I focused on handling state segregation for clean code.

manvinderjit replied:
A single update function would have become really big, making things complex and code unreadable.

madcampos replied:
That is the thing, the form doesn't need to be big and complex.

For most of the data we only need to care about it when the form is submitted. It doesn't matter what the user types in until then. We can just let the browser handle all this state for us and only care about the more complex stuff.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
-														x
-													</button>
-												</span>
-											))}
+		isLoading ?
+			<p>Loading...</p> :
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** Please prefer an early return...‏

```typescript
if (isLoading) {
    return <p>Loading...</p>;
}

return ...
```

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
+	useEffect(() => {
+		setIsLoading(true);
+		fetch(`/api/profiles/self`, {
+			method: 'GET',
+			headers: {
+				'Content-Type': 'application/json'
+			},
+			credentials: 'include'
+		})
+			.then(async (response) => {
+				// If not authenticated, redirect to Sign In page
+				if (response.status === 401) {
+					window.location.href = `/pages/sign-in/`;
+				} else if (!response.ok) {
+					throw new Error('Network response was not ok');
+				}
+				setIsLoading(false);
+				return response.json();
+			})
+			.then((data) => {
+				const fetchedFields = {
+					id: data.data?.id,
+					name: data.data?.name,
+					email: data.data?.email,
+					isBasedOnGTA: data.data?.isBasedOnGTA,
+					canJoinLocalEvents: data.data?.canJoinLocalEvents
+				};
+
+				setProfileData((prev) => ({
+					...prev,
+					...fetchedFields
+				}));
+
+				setProfileId(data.data.id);
+				setIsLoading(false);
+			})
+			.catch((_error) => {
+				// setErrorMessage(error.message);
+				setIsLoading(false);
+			});
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** Use an async [IIFE‏](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) instead of chaining `then`/`catch` calls.

```suggestion
		(async () => {
			setIsLoading(true);
			
			try {
				const response = await fetch(`/api/profiles/self`, {
					method: 'GET',
					headers: {
						'Content-Type': 'application/json'
					},
					credentials: 'include'
				});
				
				// If not authenticated, redirect to Sign In page
				if (response.status === 401) {
					window.location.href = `/pages/sign-in/`;
				} else if (!response.ok) {
					throw new Error('Network response was not ok');
				}

				const data = await response.json();
				const fetchedFields = {
					id: data.data?.id,
					name: data.data?.name,
					email: data.data?.email,
					isBasedOnGTA: data.data?.isBasedOnGTA,
					canJoinLocalEvents: data.data?.canJoinLocalEvents
				};
			
				setProfileData((prev) => ({
					...prev,
					...fetchedFields
				}));

				setProfileId(data.data.id);
			} catch (err) {
				// setErrorMessage(error.message);

			} finally {
				setIsLoading(false);
			}
		})();
```

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
+			});
+	}, []);
+
+	useEffect(() => {
+		// Update profileData only if new date is valid
+		if (isRealDate(birthdayValue)) {
+			setProfileData((prev) => ({
+				...prev,
+				birthday: birthdayValue
+			}));
+		}
+	}, [birthdayValue]);
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** ‏I think this is unnecessary as we are already updating the data inside an event handler.‏

The `useEffect` hook in react is recomended for async work or things that affect react from the outside in. User input, if we have event handlers, covers this case.

manvinderjit replied:
So I am using two states in the page
1.`birthdayValue` which stores the value of the birthday, it can have half values like `MM- `or `-DD` because it is waiting for the user to select both values. It is set using the handle function for either of the `selects` for MM or DD.
2. `profileData` is the main state that stores data fetched or sent to backend. It has a `birthday` value that stores the data to be sent to backend. It is updated using this useEffect if `birthdayValue` is valid. 

I can club both functions into the first function if you think that makes sense.

madcampos replied:
We can definitely move the date selector to a component to make it easier to handle. Also, it can do all of the required magic for disabling dates based on the month, like February disabling days 30 and 31.

manvinderjit replied:
Will move it to into a separate component once i am done with the rest of refactoring and created the input component.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
+							{ label: 'Complete your profile' }
+						]}
+					/>
+					<form onSubmit={handleSubmit} encType='multipart/form-data' id='complete-profile-form'>
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** We should add support on the server for regular form data instead of only json‏

manvinderjit replied:
Will look into this after we complete file upload.

madcampos replied:
Created an issue for that: https://github.com/torontojs/communityhub/issues/144

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
 								</div>
+							</details>
+
+							{/* TODO: Need to confirm adding green check icon on successful section completion on mobile view */}
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** Is this confirmed yet?‏

manvinderjit replied:
Every time I bring this up, the conversation gets sidetracked. :smile: 

https://torontojs.slack.com/archives/C081HT0SGMP/p1747333496124309

I am gonna skip it for now I think. :laughing:

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
 											<input
-												id={`${socialInput}-input`}
-												name={socialInput}
+												id='linkedInProfile'
+												name='linkedInProfile'
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏Please use `kebab-case` for html attribute values that are used cross languages (html, js, css)‏

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -92,327 +175,512 @@ const CompleteProfile = () => {
+		}
+	};
+
+	const linksByPlatform = useMemo(() => {
+		const map: Record<string, string> = {};
+		profileData.links.forEach(({ platform, url }) => {
+			map[platform] = url;
+		});
+		return map;
+	}, [profileData.links]);
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** this should be the way the data is represented here and only get transformed to an array when the data is submited to the server.‏

As a general tip, if the data don't quite fit the expectations of your "unit of code" (function, class, component, api endpoint, etc), transform it on input and then transform it back on the output. As long as the _contract_ between your unit of code and other units remains the same, you can do whatever with the data _inside_ your unit of code.

manvinderjit replied:
This is primarily to get the value of the `links` based on `platform` to show it in `HTML inputs`. It doesn't affect the data in the state or for the request.

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -16,30 +17,98 @@ interface SocialIcons {
+		{ name: 'Twitter', element: <XTwitter />, visible: true },
 		{ name: 'Dev.to', element: <DevTo />, visible: true }
 	]);
 	const [socialAccountInputs, setSocialAccountInputs] = useState<string[]>([]);
```
See this example of how to use inputs with the new [invoker proposal](https://open-ui.org/components/invokers.explainer/): https://codepen.io/madcampos/pen/RNPGRoQ?editors=0011

It is working on chromium based browsers but also includes a basic pollyfill for other browsers.

Once support is available across the board, we can use it (mostly) without js.

manvinderjit replied:
@madcampos  Do we implement this now or once it is supported across the board?

madcampos replied:
Up to you 🤷‍♂️

manvinderjit replied:
I will probably leave it for later, once i refactor all the code and create smaller components.

madcampos commented:
> > @manvinderjit what is the thing with cookies? Could you expand on that?
> 
> @madcampos You will have to generate a cookie by signing in to be able to access the Complete Profile page.

We had some brief conversations before about having the profile in different states/keeping flags for the steps of the profile... Maybe it is a good opportunity to revisit that.

manvinderjit commented:
> > > @manvinderjit what is the thing with cookies? Could you expand on that?
> > 
> > 
> > @madcampos You will have to generate a cookie by signing in to be able to access the Complete Profile page.
> 
> We had some brief conversations before about having the profile in different states/keeping flags for the steps of the profile... Maybe it is a good opportunity to revisit that.

Yeah, that is what is on the table. It is required for the integration that I am going to do next.

manvinderjit committed:
fix: add documentation for functions

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: QoL fixes for updateProfile function

Rename arguments from args to data.
Check if response status code validity.

manvinderjit committed:
fix: update handleAddSkill to use set object for unique values

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: add early return for loading

manvinderjit committed:
fix: replace chained if then with async IIFE

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: change input attributes to use kebab-case

manvinderjit committed:
fix: add redirect to home on success

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -1,10 +1,11 @@
 import { Facebook } from '../Icons/Social/Facebook.tsx';
 import { Instagram } from '../Icons/Social/Instagram.tsx';
-import { LinkedIn } from '../Icons/Social/LinkedIn.tsx';
+import { Mastodon } from '../Icons/Social/Mastodon.tsx';
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** did we remove linkedin and just added mastodon here?‏

manvinderjit replied:
@madcampos Negative. We have both. For LinkedIn we have a full length input. It was also being repeated in the optional account icons (which must be clicked upon to open their corresponding input) which should have been Mastodon instead.

![image](https://github.com/user-attachments/assets/ac70c01f-8cfd-4d0c-b4ff-413ebc1218d1)

manvinderjit committed:
refactor: remove state updates on user input

manvinderjit committed:
fix: change social platform enum values to lowercase

manvinderjit committed:
fix: validate slack handle upon form submission

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -1,44 +1,154 @@
+	skills: string[];
 }
 
+interface UpdateProfileParams extends Omit<ProfileParams, 'email'> {}
```
**nitpick**:
```suggestion
type UpdateProfileParams = Omit<ProfileParams, 'email'>;
```

madcampos commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -244,17 +496,18 @@ const CompleteProfile = () => {
 										/>
 										<label htmlFor='canJoinLocalEvents'>
-											<span>I can join TorontoJS's local events</span>
+											<span>I can join TorontoJS&#8217; local events</span>
```
**nitpick**:
```suggestion
											<span>I can join TorontoJS&apos; local events</span>
```

manvinderjit committed:
fix: ts type and qol fixes from code review

change typescript interface to type
change apostrophe from numeric character reference to name character entity

Co-authored-by: Marco Campos <me@madcampos.dev>

kbventures commented on api/src/routes/profile/validation.ts:
```diff
@@ -1,7 +1,7 @@
 import { BaseDbEntitySchema, BaseDBFieldsToOmit } from '../../utils/db.ts';
 
-export const PlatformEnum = z.enum(['Slack', 'LinkedIn', 'GitHub', 'Portfolio', 'Instagram', 'Threads', 'Facebook', 'BlueSky', 'Mastadon', 'Twitter', 'Dev']);
+export const PlatformEnum = z.enum(['slack', 'linkedin', 'github', 'portfolio', 'instagram', 'threads', 'facebook', 'bluesky', 'mastodon', 'xtwitter', 'dev']);
```
Question: Is this a Linter changing the casing, or is this on purpose?

manvinderjit replied:
This is on purpose, initially i had used Camel case but Marco asked this to be converted to lowercase for uniformity.

kbventures commented on src/components/CompleteProfile/CompleteProfile.css:
```diff
@@ -258,7 +258,7 @@
 		justify-content: space-between;
 
-		label:is([for='linkedInProfile'], [for='githubProfile'], [for='portfolio']) {
+		label:is([for='linkedin'], [for='github'], [for='portfolio']) {
```
Question: same as above

manvinderjit replied:
Same answer as above, for uniformity. We don't need to change fields when sending data from frontend to api.

kbventures commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -1,44 +1,154 @@
+	email: string;
+	isBasedOnGTA: boolean;
+	canJoinLocalEvents: boolean;
+	name?: string;
```
Question: Why is this optional?

manvinderjit replied:
I was going by the Figma design. It lists name as optional. But it is a readonly field and this page won't update a user's name, only the following params
![image](https://github.com/user-attachments/assets/ce6bab23-6caa-406c-a521-694e2e8a07b4)

kbventures commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -1,44 +1,154 @@
+	return date.getMonth() + 1 === month && date.getDate() === day;
+};
+
+const updateProfile = async (data: UpdateProfileParams, profileId: string) => {
```
Question: I suspect updatingProfile will surely be re-used, so why should we not move this to the hook folder?

manvinderjit replied:
Yeah, will definitely do once we have all the integration complete.

kbventures commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -73,28 +185,150 @@ const CompleteProfile = () => {
-		setSocialAccountInputs((previousAccounts) => {
-			if (!previousAccounts?.includes(socialAccountToAdd)) {
-				return [...previousAccounts, socialAccountToAdd];
+	const getProfileParams = (formData: FormData): UpdateProfileParams => {
```
Question: I suspect getProfileParams will surely be re-used, so why should we not move this to the hook folder?

manvinderjit replied:
For this we will be updating different profile params for different situations. E.g., For activation we are allowing updating of params shown in image below. It doesn't have  name, email, etc. For edit profile we will be including that. So it will be situational.
![image](https://github.com/user-attachments/assets/3e05265e-3611-4865-b081-519a9b118e2c)

kbventures commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -1,44 +1,154 @@
+				console.error(e.message);
+				console.error(e.stack);
+			} else {
+				throw new Error('Error Encountered!');
```
This error message is very generic.

manvinderjit replied:
This is only for dev mode, and it is a fallback error. Ideally it won't trigger.
![image](https://github.com/user-attachments/assets/79f90077-f74e-48fd-b155-7852aaafe621)

kbventures replied:
I mean "Error Encountered" could utilize words that help you figure out where the error is generated.

madcampos replied:
+1 here. If the error is not an instance of error, throw that, like this:
```js
if (err instanceof Error) {
  //...
} else {
  throw new Error(err);
}
```

madcampos replied:
Unrelated to that: how is the linting passing with a variable called `e`? We have a specific rule that forbids variable names lesser than 3 characters long.

manvinderjit replied:
Need to evaluate this further. Other pages are also allowing shorter error variables.
<img width="576" height="306" alt="image" src="https://github.com/user-attachments/assets/233881ba-b2fa-489d-9550-146c003a56cb" />

manvinderjit replied:
error message is now updated.

kbventures commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -73,28 +185,150 @@ const CompleteProfile = () => {
+				});
+
+				// If not authenticated, redirect to Sign In page
+				if (response.status === 401) {
```
Note: I know this is probably the way it is, because we're waiting for this issue to be handled.

https://github.com/torontojs/communityhub/issues/142

Once this PR is approved, please create an issue to utilize the heartbeat(it may need some redesign) hook to check if the person who is accessing the page has the correct access / state for this page.

manvinderjit replied:
We need to authenticate the user and have created the `profiles/self` route for it. This will be in addition to the correct access/state for this page. We will need two check at least, one for authentication other for authorization.

kbventures replied:
Please create issue(s) so we can keep track of this.

manvinderjit replied:
@kbventures The `self` route for authentication is already implemented. And authorization will be based on the state that Lyla is working on right now.

kbventures commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -1,44 +1,154 @@
+ * @param birthday - The birthday string to validate, expected in MM-DD format.
+ * @returns `true` if the format is valid, otherwise `false`.
+ */
+const isValidBirthdayFormat = (birthday: string): boolean => {
```
@madcampos @manvinderjit 

Discussion: Perhaps they don't fit the bill of utilities that belong in the utility directory, but this page is large.  Would it not make sense to move isValidBirthdayFormat, isRealDate, and updateProfile in another folder?

manvinderjit replied:
@kbventures They will be moved along with the Date Component. I am working on it already.

kbventures reviewed (approved):
@manvinderjit left some comments and discussion items, but overall, it looks great. 

Please create an issue to implement changes from https://github.com/torontojs/communityhub/issues/142 into the complete profile page once it's completed, and another to create API tests(hurl) .

manvinderjit commented on src/components/CompleteProfile/CompleteProfile.tsx:
```diff
@@ -1,44 +1,154 @@
+				console.error(e.message);
+				console.error(e.stack);
+			} else {
+				throw new Error('Error Encountered!');
```
```suggestion
				console.log(e);
				throw new Error('Unknown Error Encountered!');
```
@kbventures  I can do the above for more clarity if required.

manvinderjit committed:
fix: rename api error arg, update error msg

manvinderjit commented:
> @manvinderjit left some comments and discussion items, but overall, it looks great.
> 
> Please create an issue to implement changes from #142 into the complete profile page once it's completed, and another to create API tests(hurl) .

@kbventures Both issues are now created.

kbventures reviewed (approved):
Great work
